Ignore generic type parameter names when matching

ericstj opened the pull request:
Generic type parameter names are not significant to compat. 

Remove these from the matching key, but make the `CannotChangeParameterName` rule compare them.

ericstj committed:
Ignore generic type parameter names when matching

Generic type parameter names are not significant to compat.
Remove these from the matching key, but make the ParameterName rule
compare them.

ViktorHofer commented on src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeParameterName.cs:
```diff
@@ -11,8 +12,26 @@ namespace Microsoft.DotNet.ApiCompatibility.Rules
     /// </summary>
```
Please update the summary above

ViktorHofer commented on src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/MembersMustExist.cs:
```diff
@@ -152,7 +152,7 @@ private bool ParametersMatch(IMethodSymbol method, IMethodSymbol candidate)
 
             for (int i = 0; i < method.Parameters.Length; i++)
             {
-                if (!_settings.SymbolEqualityComparer.Equals(method.Parameters[i].Type, method.Parameters[i].Type))
```
Can you please explain this change. I was really suprised when I saw that.

ericstj replied:
I just noticed this when reviewing the code.  It looks to me like a copy-paste error when this change was introduced.  It would only be hit when searching a base type for a compatible method (as if a method was pushed down to a base type).  I think it would have excused method removals if the base type hierarchy contained an overload with the same number of parameters even if the type differed.

ericstj replied:
Added a regression test for this

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:
```diff
@@ -14,27 +14,37 @@ static SymbolExtensions()
-            symbol.ToDisplayString(s_comparisonFormat)
+        public static string ToComparisonDisplayString(this ISymbol symbol)
+        {
+            var parts = symbol.ToDisplayParts(s_comparisonFormat);
+
+            // Type parameter names are not significant. Remove these leaving behind
+            // Angle brackets and commas for each parameter > 1.
+            parts = parts.RemoveAll(p => p.Kind == SymbolDisplayPartKind.TypeParameterName);
```
When I looked at this the last time, we intentionally didn't want to use the `ToDisplayParts` API as that is documented to only exist for IDE scenarios. I think you filed an issue to have API support for omitting these parts. Can we link to that as a tracking issue with a TODO?

ericstj replied:
ToDisplayString today just calls ToDisplayParts.  @CyrusNajmabadi recommended it as well, and mentions it's commonly recommended.  I don't see any discouragement in the docs: https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.symboldisplay.todisplayparts?view=roslyn-dotnet-4.7.0

ericstj replied:
This is the issue we closed based on that recommendation.  I can link to it.

ericstj committed:
Merge branch 'main' of https://github.com/dotnet/sdk into updateApiComparisonKey

ericstj committed:
Address feedback

ericstj committed:
Merge branch 'main' into updateApiComparisonKey

ericstj committed:
Merge branch 'main' into updateApiComparisonKey

ericstj committed:
Add test for member mapping to ensure type parameters considered

Also introduce an exception instead of assert when members mapped with mismatched parameter counts.

ericstj commented:
Updating with where this is 
```
 Unexpected parameter count mismatch for 
M:System.Numerics.TotalOrderIeee754Comparer`1.GetHashCode(`0) System.Int32 System.Numerics.TotalOrderIeee754Comparer<>.GetHashCode() 1 vs
M:System.Numerics.TotalOrderIeee754Comparer`1.GetHashCode System.Int32 System.Numerics.TotalOrderIeee754Comparer<>.GetHashCode() 0
```

So my change also ends up removing the `` `0 `` as the type parameter argument in the method.  I'll need to take a closer look at the workaround in `ToComparisonDisplayString` to see if I can constrain it to only apply to parts that are not method parameters.
